drivers/led_strip/Kconfig.ws2812

```diff
@@ -30,20 +30,49 @@ config WS2812_STRIP_I2S
 
 config WS2812_STRIP_GPIO
 	bool "WS2812 LED strip GPIO driver"
-	# Only an Cortex-M0 inline assembly implementation for the nRF51
-	# is supported currently.
+	# Only an Cortex-M inline assembly implementation for the nRF91, nRF51,
+	# nRF52 and nRF53 is supported currently.
 	default y
 	depends on DT_HAS_WORLDSEMI_WS2812_GPIO_ENABLED
-	depends on SOC_SERIES_NRF51X
+	depends on (SOC_SERIES_NRF91X || SOC_SERIES_NRF51X || SOC_SERIES_NRF52X || SOC_SERIES_NRF53X)
 	select LED_STRIP_RGB_SCRATCH
 	help
-	  Enable driver for WS2812 (and compatibles) LED strip directly controlling with GPIO.
-	  The GPIO driver does bit-banging with inline assembly,
-	  and is not available on all SoCs.
+	  Enable driver for WS2812 (and compatibles) LED strip directly
+	  controlling with GPIO. The GPIO driver does bit-banging with inline
+	  assembly, and is not available on all SoCs.
 
 	  Note that this driver is not compatible with the Everlight B1414
 	  controller.
 
+if WS2812_STRIP_GPIO
+
+config DELAY_T1H
+	int
+	default $(div,$(mul,750,$(dt_node_int_prop_int,/cpus/cpu@0,clock-frequency)),1000000000)
```

Comment:
> I would prefer to store delays in nanoseconds with these Kconfig options and let the driver do the conversion to number of nops. It would be nice to spare users these calculations :)

It is not possible by LISTIFY limitation.
The macro accepts only calculated macro, and cannot calculate in C.

Reply:
That's correct, the driver implements delays by adding (compile time) 'nop' asm instructions. The LISTIFY macro (which does this) can only do this with constant values. @soburi did a great job making this even possible by adding the option to use arithmetic functions to Kconfig files. Currently I don't see any other way of doing this.

Reply:
OK, then maybe we can have in Kconfig `DELAY_T{0,1}{L,H}_NS` configured by user with a value in nanosecond and `DELAY_T{0,1}{L,H}_NOP` hidden, with a number of nops and selected/computed automatically ?

Reply:
I don't think this will be a problem. I'll have a look.

Reply:
@soburi I had a look into calculating delays from nanoseconds to nops using Kconfig files, but I seem to be running to a dead end. When I try to create a Kconfig file the resulting argument to 'mul' or 'div' always seems to be empty when I base this on a Kconfig symbol, and not a constant value.

The kconfiglib library doesn't seem to support this use-case. Have a look at the warning in the documentation in the section "Preprocessor user functions defined in Python": https://github.com/ulfalizer/Kconfiglib/blob/master/kconfiglib.py#L528

```
Warning
*******

User-defined preprocessor functions are called as they're encountered at parse
time, before all Kconfig files have been processed, and before the menu tree
has been finalized. There are no guarantees that accessing Kconfig symbols or
the menu tree via the 'kconf' parameter will work, and it could potentially
lead to a crash.

Preferably, user-defined functions should be stateless.
```
What do you think?

Reply:
@chaim-zax @simonguinot 

Basically, zephyr's macro is not "everything works well".
It carries over the bad points of C language macros.
Since, this is a limitation of the current Zephyr system.

I consider the current implementation to be the limit of improvement at this point (Still, it's become pretty good, right?).

In cases like this, it's a good idea to provide sample, concrete, and easy-to-apply examples in the documentation.

Reply:
Another answer.

Add frequency property to "worldsemi,ws2812-gpio".

Reply:
Indeed, there is still an option to move some of these properties to the device tree. This would allow to do the calculations from nanoseconds to nop's as well (not the most pretty option, using both dts and Kconfig files), and possibly have a frequency option there too. @simonguinot what's your take on this?

Reply:
> I don't think this will be a problem. I'll have a look. Although the cpu clock-frequency in the dts tree might not always be available.

Note that you can use a second `default` as a fallback if the clock-frequency property is missing. For example:

```
default $(div,$(mul,750,$(dt_node_int_prop_int,/cpus/cpu@0,clock-frequency)),1000000000) \
            if $(dt_node_has_prop,/cpus/cpu@0,clock-frequency)
default 12
```

Reply:
> Indeed, there is still an option to move some of these properties to the device tree. This would allow to do the calculations from nanoseconds to nop's as well (not the most pretty option, using both dts and Kconfig files), and possibly have a frequency option there too. @simonguinot what's your take on this?

I understand the limitation. So we can't offer a nanosecond configuration to the user but we have to stick to NOPs.

Here are my suggestions:

- we add new properties to the `ws2812-gpio` DT bindings to allow the user to configure the number of NOPs for the T0L, T0H, T1L and T1H periods. There is no reason for not offering DT configuration like the other ws2812 bindings.
- we write a clear and user friendly description for these new properties. So the user will be able to configure them wisely :)
- we add the corresponding Kconfig options (as in your patch), with the first `default` set to the DT value, the second to a value calculated based on the `clock-frequency` property (if available) and the ws2812 timings (as found in the specification) and a third `default` fallback set to an hardcoded value (also wisely chosen). And we allow the user to override these values ​​from their own configuration.

How does that sound ? :)

Reply:
Sounds good, thanks for the feedback!

@soburi do you have any remarks on this, shall I just go ahead with the implementation as described here?